### PR TITLE
Deduplicate Unbounded Reads

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This ensures that sources that produce duplicate elements that are
marked as requiresDeduplication are handled by the DirectRunner.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadDeduplicator.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    private final LoadingCache<StructuralKey<byte[]>, AtomicBoolean> ids;
+
+    /**
+     * Create a new {@link CachedIdDeduplicator}.
+     */
+    public static UnboundedReadDeduplicator create() {
+      return new CachedIdDeduplicator();
+    }
+
+    private  CachedIdDeduplicator() {
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadDeduplicator.java]: why 2 spaces?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactoryTest.java [hunk 2]
```diff
@@ -332,7 +371,15 @@ public T getCurrent() throws NoSuchElementException {
 
       @Override
       public Instant getCurrentTimestamp() throws NoSuchElementException {
-        return Instant.now();
+        return new Instant(index);
+      }
+
+      public byte[] getCurrentRecordId() {
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactoryTest.java]: `@Override`?

**Comment by dhalperi** [comment]
LGTM

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactoryTest.java (hunk 2 above)
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactoryTest.java]: Done.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadDeduplicator.java (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadDeduplicator.java]: I hit the spacebar twice.

(Fixed.)

**Pull request closed without merging**